lib/parse-dir.js

```diff
@@ -1,168 +1,183 @@
-
-    var handleAsync = function(pattern, callback) {
-      that.data = [];
-      that.__assync = typeof callback === 'function';
-
-      if (that.__assync) {
-        glob(pattern, function(err, files) {
-          if (err) {
-            callback(err);
-            return;
-          }
-
-          async.each(files, that.parseFile.bind(that), function(err) {
-            callback(err, that.data);
-          });
-        });
-
-        return;
+      if (!error) {
+        return done(null, stats.isDirectory());
       }
-
-      return sync(pattern);
+      return done(error);
     };
-    handleAsync.sync = sync.bind(that);
-
-    return handleAsync;
-  };
-
-  ParseDir.prototype.parseFile = function(file, callback) {
-    var meta = {
-      relativePath: file,
-      filepath: path.resolve(file),
-      filename: path.basename(file),
-      basename: path.basename(file, path.extname(file)),
-      extension: path.extname(file),
-      parsed: false
+    if (results.__async) {
+      fs.lstat(results.meta.filepath, onLsStat);
+    } else {
+      onLsStat(null, fs.lstatSync(results.meta.filepath));
+    }
+  }
+  __getFileContents (done, results) {
+    // Get the files raw contents
+    if (results.isDirectory) {
+      return done();
+    }
+    const onFileRead = (err, fileData) => {
+      let error = null;
+      if (err || !fileData) {
+        error = err || true;
+      }
+      if (!error) {
+        return done(null, fileData.toString());
+      }
+      done(error);
     };
-    var that = this;
-
-    async.series([
-      function(done) {
-        var onLsStat = function (err, stats) {
-          var error = null;
-
-          if (err || !stats) {
-            error = err || true;
-          }
-
-          if (!error) {
-            meta.isDirectory = stats.isDirectory();
-          }
-
-          done(error);
-        };
-
-        if (that.__async) {
-          fs.lstat(meta.filepath, onLsStat);
-        } else {
-          onLsStat(null, fs.lstatSync(meta.filepath));
-        }
+    if (results.__async) {
+      fs.readFile(results.meta.filepath, onFileRead);
+    } else {
+      onFileRead(null, fs.readFileSync(results.meta.filepath));
+    }
+  }
+  __parseFileContents (done, results) {
+    // Parse the file using the correct lib
+    const meta = results.meta;
+    if (!results.fileContents) {
+      return done();
+    }
+    meta.raw = results.fileContents;
+    switch (meta.extension) {
+      case '.json':
+        jsonParser(meta, done);
+        break;
+      case '.yml':
+      case '.yaml':
+        yamlParser(meta, done);
+        break;
+      case '.coffee':
+      case '.litcoffee':
+      case '.ls':
+      case '.livescript':
+      case '.js':
+        jsParser(meta, done);
+        break;
+      default:
+        return done();
+    }
+  }
+
+  processFile (file, callback) {
+    const isAsync = this.__async;
+
+    async.auto({
+      __async: (done) => {
+        return done(null, isAsync);
       },
```

Comment:
We could leverage the auto return here and just below.

`__async: done => done(null, isAsync),`